File: travel-planning/backend/app/models/trip.py
"""Trip model."""

import uuid
from datetime import datetime

from sqlalchemy import (
    Boolean,
    DateTime,
    Enum,
    Float,
    Index,
    Integer,
    String,
)
from sqlalchemy.dialects.postgresql import JSONB, UUID
from sqlalchemy.orm import Mapped, mapped_column, relationship
from sqlalchemy.schema import ForeignKey
from sqlalchemy.sql import func

from app.core.database import Base
from app.models.enums import TravelMode, TripStatus


class Trip(Base):
    __tablename__ = "trips"
    __table_args__ = (
        Index(
            "ix_trips_user_status_arrival",
            "user_id",
            "status",
            "arrival_time",
        ),
    )

    id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), primary_key=True, default=uuid.uuid4
    )
    user_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True),
        ForeignKey("users.id", ondelete="CASCADE"),
        nullable=False,
        index=True,
    )
    name: Mapped[str] = mapped_column(String(255), nullable=False)

    # Origin
    origin_location_id: Mapped[uuid.UUID | None] = mapped_column(
        UUID(as_uuid=True),
        ForeignKey("saved_locations.id", ondelete="SET NULL"),
        nullable=True,
    )
    origin_address: Mapped[str] = mapped_column(
        String(500), nullable=False, default=""
    )
    origin_lat: Mapped[float] = mapped_column(
        Float, nullable=False, default=0.0
    )
    origin_lng: Mapped[float] = mapped_column(
        Float, nullable=False, default=0.0
    )
    origin_is_current_location: Mapped[bool] = mapped_column(
        Boolean, nullable=False, default=False
    )

    # Destination
    dest_location_id: Mapped[uuid.UUID | None] = mapped_column(
        UUID(as_uuid=True),
        ForeignKey("saved_locations.id", ondelete="SET NULL"),
        nullable=True,
    )
    dest_address: Mapped[str] = mapped_column(String(500), nullable=False)
    dest_lat: Mapped[float] = mapped_column(Float, nullable=False)
    dest_lng: Mapped[float] = mapped_column(Float, nullable=False)

    # Timing
    arrival_time: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), nullable=False, index=True
    )
    travel_mode: Mapped[TravelMode] = mapped_column(
        Enum(TravelMode, name="travelmode", create_constraint=True),
        nullable=False,
        default=TravelMode.driving,
    )
    buffer_minutes: Mapped[int] = mapped_column(
        Integer, nullable=False, default=15
    )

    # Monitoring state
    status: Mapped[TripStatus] = mapped_column(
        Enum(TripStatus, name="tripstatus", create_constraint=True),
        nullable=False,
        default=TripStatus.pending,
        index=True,
    )
    monitoring_started_at: Mapped[datetime | None] = mapped_column(
        DateTime(timezone=True), nullable=True
    )
    last_eta_seconds: Mapped[int | None] = mapped_column(
        Integer, nullable=True
    )
    last_checked_at: Mapped[datetime | None] = mapped_column(
        DateTime(timezone=True), nullable=True
    )
    notify_at: Mapped[datetime | None] = mapped_column(
        DateTime(timezone=True), nullable=True, index=True
    )
    notified: Mapped[bool] = mapped_column(
        Boolean, nullable=False, default=False
    )
    notification_count: Mapped[int] = mapped_column(
        Integer, nullable=False, default=0
    )
    baseline_duration_seconds: Mapped[int | None] = mapped_column(
        Integer, nullable=True
    )

    # Recurrence
    is_recurring: Mapped[bool] = mapped_column(
        Boolean, nullable=False, default=False
    )
    recurrence_rule: Mapped[dict | None] = mapped_column(
        JSONB, nullable=True
    )

    # Calendar integration
    calendar_event_id: Mapped[str | None] = mapped_column(
        String(500), nullable=True, index=True
    )

    # Soft delete
    is_deleted: Mapped[bool] = mapped_column(
        Boolean, nullable=False, default=False
    )

    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), nullable=False, server_default=func.now()
    )
    updated_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True),
        nullable=False,
        server_default=func.now(),
        onupdate=func.now(),
    )

    # Relationships
    user = relationship("User", back_populates="trips")
    origin_location = relationship(
        "SavedLocation", foreign_keys=[origin_location_id]
    )
    dest_location = relationship(
        "SavedLocation", foreign_keys=[dest_location_id]
    )
    eta_snapshots = relationship(
        "TripEtaSnapshot",
        back_populates="trip",
        cascade="all, delete-orphan",
    )
    notifications = relationship(
        "NotificationLog",
        back_populates="trip",
        cascade="all, delete-orphan",
    )

File: travel-planning/ios/Depart/Depart/Models/Trip.swift
import Foundation

// MARK: - Enums

/// Matches backend `app/models/enums.py` TravelMode exactly.
enum TravelMode: String, Codable, CaseIterable {
    case driving
    case transit
    case walking
    case cycling
}

/// Matches backend `app/models/enums.py` TripStatus exactly.
enum TripStatus: String, Codable {
    case pending
    case monitoring
    case notified
    case departed
    case completed
    case cancelled
}

/// Matches backend `app/models/enums.py` CongestionLevel exactly.
enum CongestionLevel: String, Codable {
    case unknown
    case light
    case moderate
    case heavy
    case severe

    var displayName: String {
        rawValue.capitalized
    }
}

// MARK: - Trip Model

/// Matches backend `app/schemas/trip.py` TripResponse exactly.
/// Uses convertFromSnakeCase key strategy on the decoder.
struct Trip: Identifiable, Codable, Hashable {
    let id: UUID
    var name: String

    // Origin
    var originAddress: String
    var originLat: Double
    var originLng: Double
    var originLocationId: UUID?
    var originIsCurrentLocation: Bool

    // Destination
    var destAddress: String
    var destLat: Double
    var destLng: Double
    var destLocationId: UUID?

    // Timing
    var arrivalTime: Date
    var travelMode: String // Raw string from backend (e.g., "driving")
    var bufferMinutes: Int

    // Monitoring state
    var status: String // Raw string from backend (e.g., "pending")
    var monitoringStartedAt: Date?
    var lastEtaSeconds: Int? // "lastEtaSeconds" matches convertFromSnakeCase("last_eta_seconds")
    var lastCheckedAt: Date?
    var notifyAt: Date?
    var baselineDurationSeconds: Int?
    var notified: Bool
    var notificationCount: Int

    // Recurrence
    var isRecurring: Bool
    var recurrenceRule: [String: AnyCodableValue]?

    // Calendar
    var calendarEventId: String?

    // Timestamps
    var createdAt: Date
    var updatedAt: Date
}

// MARK: - Trip Computed Properties

extension Trip {
    var travelModeEnum: TravelMode {
        TravelMode(rawValue: travelMode) ?? .driving
    }

    var statusEnum: TripStatus {
        TripStatus(rawValue: status) ?? .pending
    }

    /// Recommended departure time based on server-computed notifyAt.
    var departureTime: Date? {
        notifyAt
    }

    /// Minutes until recommended departure. Negative if overdue.
    var minutesUntilDeparture: Int? {
        guard let departureTime else { return nil }
        return Int(departureTime.timeIntervalSinceNow / 60)
    }

    /// Estimated travel time in minutes.
    var estimatedTravelMinutes: Int? {
        guard let lastEtaSeconds else { return nil }
        return lastEtaSeconds / 60
    }

    /// Status color category for dashboard display.
    var urgencyLevel: UrgencyLevel {
        guard let minutes = minutesUntilDeparture else { return .normal }
        if minutes < 0 { return .overdue }
        if minutes < 10 { return .critical }
        if minutes < 30 { return .warning }
        return .normal
    }

    enum UrgencyLevel {
        case normal   // > 30 min — green
        case warning  // 10-30 min — yellow
        case critical // < 10 min — red
        case overdue  // past departure — red
    }
}

// MARK: - AnyCodableValue

/// Lightweight wrapper for encoding/decoding untyped JSON values (for recurrence_rule JSONB).
enum AnyCodableValue: Codable, Hashable {
    case string(String)
    case int(Int)
    case double(Double)
    case bool(Bool)
    case array([AnyCodableValue])
    case dictionary([String: AnyCodableValue])
    case null

    init(from decoder: Decoder) throws {
        let container = try decoder.singleValueContainer()
        if container.decodeNil() {
            self = .null
        } else if let value = try? container.decode(Bool.self) {
            self = .bool(value)
        } else if let value = try? container.decode(Int.self) {
            self = .int(value)
        } else if let value = try? container.decode(Double.self) {
            self = .double(value)
        } else if let value = try? container.decode(String.self) {
            self = .string(value)
        } else if let value = try? container.decode([AnyCodableValue].self) {
            self = .array(value)
        } else if let value = try? container.decode([String: AnyCodableValue].self) {
            self = .dictionary(value)
        } else {
            self = .null
        }
    }

    func encode(to encoder: Encoder) throws {
        var container = encoder.singleValueContainer()
        switch self {
        case .string(let v): try container.encode(v)
        case .int(let v): try container.encode(v)
        case .double(let v): try container.encode(v)
        case .bool(let v): try container.encode(v)
        case .array(let v): try container.encode(v)
        case .dictionary(let v): try container.encode(v)
        case .null: try container.encodeNil()
        }
    }
}
